File: digitalpassbook2/app/src/main/java/com/example/digitalpassbook/StudentService.java
package com.example.digitalpassbook;

import java.util.List;
import retrofit2.Call;
import retrofit2.http.Body;
import retrofit2.http.GET;
import retrofit2.http.POST;
import retrofit2.http.PUT;
import retrofit2.http.PATCH;
import retrofit2.http.DELETE;
import retrofit2.http.Path;

public interface StudentService {
    @GET ("students")
    Call<List<Student>> getall();

    @POST ("students/new")
    Call<Student> create(@Body Student student);

    @GET ("students/{id}")
    Call<Student> get(@Path("id") int id);

    @DELETE ("students/{id}")
    Call<Student> delete(@Path("id") int id);

    @PATCH ("students/{id}")
    Call<Student> update(@Path("id") int id, @Body Student student);

}

File: digitalpassbook2/app/src/main/java/com/example/digitalpassbook2/StudentService.kt
package com.example.digitalpassbook2

import retrofit2.Call
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import retrofit2.converter.scalars.ScalarsConverterFactory
import retrofit2.http.*

interface StudentService {
    @GET("students")
    fun getall(): Call<List<Student?>?>?

    @POST("students/new")
    fun create(@Body student: Student?): Call<Student?>?

    @GET("students/{id}")
    operator fun get(@Path("id") id: Int): Call<Student?>?

    @DELETE("students/{id}")
    fun delete(@Path("id") id: Int): Call<Student?>?

    @PATCH("students/{id}")
    fun update(
        @Path("id") id: Int,
        @Body student: Student?
    ): Call<Student?>?

    companion object {
        fun create(): StudentService {

            val retrofit = Retrofit.Builder()
                .baseUrl("https://pure-river-68629.herokuapp.com/")
                .addConverterFactory(ScalarsConverterFactory.create())
                .addConverterFactory(GsonConverterFactory.create())
                .build();


            return retrofit.create(StudentService::class.java)
        }
    }
}